core/src/main/java/org/elasticsearch/discovery/zen/fd/NodesFaultDetection.java

```diff
@@ -284,6 +282,16 @@ public void messageReceived(PingRequest request, TransportChannel channel) throw
                     + clusterName + "]");
             }
 
+            final DiscoveryNode currentMasterNode = clusterService.state().nodes().getMasterNode();
+            if (currentMasterNode != null
+                    && request.masterNode != null
+                    && currentMasterNode.equals(request.masterNode) == false) {
+                // this node has a different node as master than the one that sent this ping message, so notify
+                // the sending node that this node does not believe it to be the master
+                throw new IllegalStateException("Got pinged by node [" + request.masterNode + "] but I believe master to be ["
```

Comment:
I believe we should actually retry if this situation is met, because its possible that the current node has not yet processed a new cluster state with a new master, so we should retry waiting for it to get the new cluster state and correctly respond to the ping.  I believe the other two situations above that throw an IllegalStateException should stop retries from occurring.  What do you think @bleskes ?